macros/protobuf-derive/src/lib.rs

```diff
@@ -0,0 +1,119 @@
+use darling::{util::Flag, FromMeta};
+use proc_macro::TokenStream;
+use syn::{parse_macro_input, spanned::Spanned};
+
+mod proto;
+mod raw;
+
+#[proc_macro_derive(Protobuf, attributes(proto))]
+pub fn proto_derive(input: TokenStream) -> TokenStream {
+    proto::expand_raw_existing(parse_macro_input!(input))
+        .unwrap_or_else(syn::Error::into_compile_error)
+        .into()
+}
+
+#[proc_macro_derive(Raw, attributes(raw))]
+pub fn raw_derive(input: TokenStream) -> TokenStream {
+    raw::extend_new_structure(parse_macro_input!(input))
+        .unwrap_or_else(syn::Error::into_compile_error)
+        .into()
+}
+
+#[derive(FromMeta, Default)]
+#[darling(and_then = Self::not_both)]
+struct OptionalOrRepeated {
+    optional: Flag,
+    repeated: Flag,
+}
+
+impl OptionalOrRepeated {
+    fn not_both(self) -> darling::Result<Self> {
+        if self.optional.is_present() && self.repeated.is_present() {
+            Err(
+                darling::Error::custom("Cannot set `optional` and `repeated`")
```

Comment:
I'm not sure about this case. It seems like can exist field `Option<Vec<msg>>`. 

For now we can save this restriction.

Reply:
Technically, it is possible even now.  `#[raw(raw = Vec::<T>, optional)]`